File: site/src/lib/api/body_data.ts
/**
 * @file body_data.ts
 * @brief Body Data (身体数据) API
 * @author sailing-innocent
 * @date 2026-09-06
 *
 * 对应后端 BodyDataController（/api/v1/health/body-data）。
 * 核心语义：data JSON 中不存在某 key = 本次未测量。
 */

import { SERVER_URL, API_BASE } from './config'
import {
  type BodyDataRecord,
  type BodyDataCreateProps,
  type BodyDataUpdateProps,
  type BodyDataSeriesResponse,
  type BodyMetricSchemaResponse,
  type BodyMetricAnalysisResult,
} from '@lib/data/body_data'

const BODY_DATA_API_BASE = API_BASE + '/health/body-data'

// ==================== Body Data CRUD APIs ====================

const api_create_body_data = async (props: BodyDataCreateProps): Promise<BodyDataRecord> => {
  try {
    const response = await fetch(`${SERVER_URL}/${BODY_DATA_API_BASE}/`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(props),
    })
    if (!response.ok) {
      throw new Error(`Error creating body data: ${response.statusText}`)
    }
    return response.json()
  } catch (error) {
    console.error('Failed to create body data:', error)
    throw error
  }
}

const api_get_body_data_list = async (
  skip: number = 0,
  limit: number = -1,
  start: number = -1,
  end: number = -1,
  metric: string = ''
): Promise<BodyDataRecord[]> => {
  try {
    const params = new URLSearchParams()
    params.append('skip', skip.toString())
    params.append('limit', limit.toString())
    params.append('start', start.toString())
    params.append('end', end.toString())
    if (metric) {
      params.append('metric', metric)
    }
    const response = await fetch(`${SERVER_URL}/${BODY_DATA_API_BASE}/?${params}`)
    if (!response.ok) {
      throw new Error(`Error fetching body data list: ${response.statusText}`)
    }
    return response.json()
  } catch (error) {
    console.error('Failed to fetch body data list:', error)
    throw error
  }
}

const api_get_body_data = async (id: number): Promise<BodyDataRecord> => {
  try {
    const response = await fetch(`${SERVER_URL}/${BODY_DATA_API_BASE}/${id}`)
    if (!response.ok) {
      throw new Error(`Error fetching body data: ${response.statusText}`)
    }
    return response.json()
  } catch (error) {
    console.error('Failed to fetch body data:', error)
    throw error
  }
}

const api_update_body_data = async (
  id: number,
  props: BodyDataUpdateProps
): Promise<BodyDataRecord> => {
  try {
    const response = await fetch(`${SERVER_URL}/${BODY_DATA_API_BASE}/${id}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(props),
    })
    if (!response.ok) {
      throw new Error(`Error updating body data: ${response.statusText}`)
    }
    return response.json()
  } catch (error) {
    console.error('Failed to update body data:', error)
    throw error
  }
}

const api_delete_body_data = async (id: number): Promise<void> => {
  try {
    const response = await fetch(`${SERVER_URL}/${BODY_DATA_API_BASE}/${id}`, {
      method: 'DELETE',
    })
    if (!response.ok) {
      throw new Error(`Error deleting body data: ${response.statusText}`)
    }
  } catch (error) {
    console.error('Failed to delete body data:', error)
    throw error
  }
}

// ==================== Body Metric Schema / Series / Analysis APIs ====================

const api_get_body_metrics = async (): Promise<BodyMetricSchemaResponse> => {
  try {
    const response = await fetch(`${SERVER_URL}/${BODY_DATA_API_BASE}/metrics`)
    if (!response.ok) {
      throw new Error(`Error fetching body metrics: ${response.statusText}`)
    }
    return response.json()
  } catch (error) {
    console.error('Failed to fetch body metrics:', error)
    throw error
  }
}

const api_get_body_series = async (
  metric: string,
  start: number = -1,
  end: number = -1
): Promise<BodyDataSeriesResponse> => {
  try {
    const params = new URLSearchParams()
    params.append('metric', metric)
    if (start > 0) params.append('start', start.toString())
    if (end > 0) params.append('end', end.toString())
    const response = await fetch(`${SERVER_URL}/${BODY_DATA_API_BASE}/series?${params}`)
    if (!response.ok) {
      throw new Error(`Error fetching body series: ${response.statusText}`)
    }
    return response.json()
  } catch (error) {
    console.error('Failed to fetch body series:', error)
    throw error
  }
}

const api_analyze_body_metric = async (
  metric: string,
  modelType: string = 'linear',
  start: number = -1,
  end: number = -1
): Promise<BodyMetricAnalysisResult> => {
  try {
    const params = new URLSearchParams()
    params.append('metric', metric)
    params.append('model_type', modelType)
    if (start > 0) params.append('start', start.toString())
    if (end > 0) params.append('end', end.toString())
    const response = await fetch(`${SERVER_URL}/${BODY_DATA_API_BASE}/analysis?${params}`)
    if (!response.ok) {
      throw new Error(`Error analyzing body metric: ${response.statusText}`)
    }
    return response.json()
  } catch (error) {
    console.error('Failed to analyze body metric:', error)
    throw error
  }
}

// prettier-ignore
export {
  api_create_body_data,
  api_get_body_data_list,
  api_get_body_data,
  api_update_body_data,
  api_delete_body_data,
  api_get_body_metrics,
  api_get_body_series,
  api_analyze_body_metric,
}

File: sail_server/controller/body_data.py
# -*- coding: utf-8 -*-
# @file body_data.py
# @brief Body Data Controller
# @author sailing-innocent
# @date 2026-09-06
# @version 1.0
# ---------------------------------

"""
身体数据控制器。

API 契约（前缀 /api/v1/health/body-data）：
    POST   /body-data            创建记录（data 至少含 1 个指标；含 weight 时 dual-write）
    GET    /body-data            列表（skip/limit/start/end/metric 过滤）
    GET    /body-data/metrics    指标注册表（内置 + 自定义扫描）
    GET    /body-data/series     单指标时序（?metric=&start=&end=）
    GET    /body-data/analysis   单指标趋势分析（?metric=&start=&end=&model_type=）
    GET    /body-data/{id}       单条记录
    PUT    /body-data/{id}       更新记录（weight 联动更新/补建/删除）
    DELETE /body-data/{id}       删除记录（级联 dual-written Weight 行）

校验失败（data 为空 / 非法 key / 超范围 value）返回 422 及合法 key 列表。
"""
from __future__ import annotations

import logging
from typing import Generator, List

from litestar import Controller, Request, Response, delete, get, post, put
from litestar.exceptions import ValidationException
from litestar.status_codes import HTTP_422_UNPROCESSABLE_ENTITY
from sqlalchemy.orm import Session

from sail_server.application.dto.body_data import (
    BodyDataCreateRequest,
    BodyDataResponse,
    BodyDataSeriesResponse,
    BodyDataUpdateRequest,
    BodyMetricDefinition,
)
from sail_server.model.body_data import (
    analyze_body_metric_trend_impl,
    create_body_data_impl,
    delete_body_data_impl,
    get_metric_schema_impl,
    get_metric_series_impl,
    read_body_data_impl,
    read_body_data_list_impl,
    update_body_data_impl,
)

logger = logging.getLogger(__name__)


def _validation_to_422(request: Request, exc: ValidationException) -> Response:
    """将 pydantic 请求校验失败映射为 422（附合法 key 列表，见计划 2.5 节）。"""
    return Response(
        content={
            "status_code": HTTP_422_UNPROCESSABLE_ENTITY,
            "detail": exc.detail,
            "extra": exc.extra,
        },
        status_code=HTTP_422_UNPROCESSABLE_ENTITY,
    )


class BodyDataController(Controller):
    path = "/body-data"

    exception_handlers = {ValidationException: _validation_to_422}

    # POST /body-data
    @post()
    async def create_body_data(
        self,
        data: BodyDataCreateRequest,
        request: Request,
        router_dependency: Generator[Session, None, None],
    ) -> BodyDataResponse:
        """Create a new body data record."""
        db = next(router_dependency)
        record = create_body_data_impl(db, data)
        logger.info(f"[body_data] create #{record.id}: {list(record.data.keys())}")
        return record

    # GET /body-data?skip=&limit=&start=&end=&metric=
    @get()
    async def get_body_data_list(
        self,
        router_dependency: Generator[Session, None, None],
        skip: int = 0,
        limit: int = 10,
        start: float | None = None,  # timestamp as float in seconds
        end: float | None = None,
        metric: str | None = None,
    ) -> List[BodyDataResponse]:
        """List body data records with optional metric filter."""
        db = next(router_dependency)
        records = read_body_data_list_impl(db, skip, limit, start, end, metric)
        return records

    # GET /body-data/metrics
    @get("/metrics")
    async def get_metrics(
        self,
        router_dependency: Generator[Session, None, None],
    ) -> List[BodyMetricDefinition]:
        """Get metric registry (builtin + custom metrics discovered from data)."""
        db = next(router_dependency)
        return get_metric_schema_impl(db)

    # GET /body-data/series?metric=&start=&end=
    @get("/series")
    async def get_metric_series(
        self,
        router_dependency: Generator[Session, None, None],
        metric: str,
        start: float | None = None,
        end: float | None = None,
    ) -> BodyDataSeriesResponse:
        """Get single-metric time series (missing measurements are omitted)."""
        db = next(router_dependency)
        return get_metric_series_impl(db, metric, start, end)

    # GET /body-data/analysis?metric=&start=&end=&model_type=
    @get("/analysis")
    async def analyze_metric_trend(
        self,
        router_dependency: Generator[Session, None, None],
        metric: str,
        start: float | None = None,
        end: float | None = None,
        model_type: str = "linear",  # 'linear' or 'polynomial'
    ) -> dict:
        """Analyze trend for any body metric (same regression core as weight)."""
        db = next(router_dependency)
        result = analyze_body_metric_trend_impl(db, metric, start, end, model_type)
        logger.info(
            f"[body_data] analysis metric={metric} slope={result['slope']}, "
            f"trend={result['current_trend']}"
        )
        return result

    # GET /body-data/{record_id:int}
    @get("/{record_id:int}")
    async def get_body_data(
        self,
        record_id: int,
        router_dependency: Generator[Session, None, None],
        request: Request,
    ) -> BodyDataResponse | None:
        """Get a single body data record."""
        db = next(router_dependency)
        return read_body_data_impl(db, record_id)

    # PUT /body-data/{record_id:int}
    @put("/{record_id:int}")
    async def update_body_data(
        self,
        record_id: int,
        data: BodyDataUpdateRequest,
        request: Request,
        router_dependency: Generator[Session, None, None],
    ) -> BodyDataResponse | None:
        """Update a body data record (syncs dual-written Weight row)."""
        db = next(router_dependency)
        record = update_body_data_impl(db, record_id, data)
        logger.info(f"[body_data] update #{record_id}: {record is not None}")
        return record

    # DELETE /body-data/{record_id:int}
    @delete("/{record_id:int}", status_code=200)
    async def delete_body_data(
        self,
        record_id: int,
        request: Request,
        router_dependency: Generator[Session, None, None],
    ) -> dict:
        """Delete a body data record (cascades to dual-written Weight row)."""
        db = next(router_dependency)
        ok = delete_body_data_impl(db, record_id)
        logger.info(f"[body_data] delete #{record_id}: {ok}")
        return {"deleted": ok, "id": record_id}
